tools/deploy-cli/deploy-cli.go

```diff
@@ -0,0 +1,324 @@
+package main
+
+import (
+	"context"
+	"crypto/rand"
+	"encoding/base64"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+
+	"embed"
+
+	"golang.org/x/crypto/bcrypt"
+	"net/http"
+	"sigs.k8s.io/cluster-api/test/framework"
+	testexec "sigs.k8s.io/cluster-api/test/framework/exec"
+	"text/template"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/kustomize/api/krusty"
+	"sigs.k8s.io/kustomize/kyaml/filesys"
+)
+
+// DeployContext defines the context of the deploy run
+type DeployContext struct {
+	// Whether to deploy with basic auth
+	DeployBasicAuth bool
+	// Whether to deploy with TLS
+	DeployTLS bool
+	// Whether to deploy KeepAlived
+	DeployKeepAlived bool
+	// Whether to deploy Mariadb
+	DeployMariadb bool
+	// string represents whether to deploy Ironic with RestartContainerCertificateUpdated
+	RestartContainerCertificateUpdated string
+	// Endpoint for Ironic
+	IronicHostIP string
+	// Endpoint for Mariadb
+	MariaDBHostIP string
+	// Templates to render files using in deployments
+	TemplateFiles embed.FS
+}
+
+// GetEnvOrDefault returns the value of the environment variable key if it exists
+// and is non-empty. Otherwise it returns the provided default value.
+func GetEnvOrDefault(key, defaultValue string) string {
+	value, exists := os.LookupEnv(key)
+	if exists && value != "" {
+		return value
+	}
+
+	return defaultValue
+}
+
+// GenerateHtpasswd generates a htpasswd entry for the given username and password.
+func GenerateHtpasswd(username, password string) (string, error) {
+	hashedPassword, err := bcrypt.GenerateFromPassword([]byte(password), bcrypt.DefaultCost)
+	if err != nil {
+		return "", err
+	}
+
+	return fmt.Sprintf("%s:%s", username, string(hashedPassword)), nil
+}
+
+// GenerateRandomString generates random string of given length
+// using crypto/rand and base64 encoding.
+func GenerateRandomString(length int) (string, error) {
+	b := make([]byte, length)
+	_, err := rand.Read(b)
+	if err != nil {
+		return "", fmt.Errorf("failed to generate random string: %v", err)
+	}
+
+	return base64.RawURLEncoding.EncodeToString(b)[:length], nil
+}
+
+// DownloadFile downloads a file and stores its content to a specified location on disk
+func DownloadFile(url string, filepath string) error {
+	// Create the file
+	out, err := os.Create(filepath)
+	if err != nil {
+		return err
+	}
+	defer out.Close()
+
+	// Get the data
+	resp, err := http.Get(url)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	// Write the body to file
+	_, err = io.Copy(out, resp.Body)
+	return err
+}
+
+// EnsureFileExists checks if a file exists at path, creating it with random content if it doesn't.
+func EnsureFileExists(varName, path string, length int) (bool, error) {
```

Comment:
I would be extremely surprised as a user of this function. Ensuring that the file exists should not imply that it is filled with random data. This should be called something else (maybe `EnsureCredentials`) or perhaps rather split up in two: ensure the file is there and ensure that it has some random content.
We should probably also validate the content when the file does exist.
Finally the input seems a bit random. `varName` is just used in a log message.

Reply:
Thanks. I have not gone through this one very carefully, I think we may not even need it, I'll check.